crates/holochain_state/src/source_chain.rs

```diff
@@ -1652,47 +1651,42 @@ pub mod tests {
             let action = chain
                 .put_weightless(action_builder, Some(entry), ChainTopOrdering::default())
                 .await?;
-
             chain.flush(&mock).await.unwrap();
-
             (action, entry_hash)
         };
 
-        {
-            let chain = SourceChain::new(
-                db.clone(),
-                dht_db.to_db(),
-                dht_db_cache.clone(),
-                keystore.clone(),
-                alice.clone(),
-            )
-            .await?;
-            // alice should find her own authorship with higher priority than the committed grant
-            // even if she passes in the secret
-            assert_eq!(
-                chain
-                    .valid_cap_grant(function.clone(), alice.clone(), secret.clone())
-                    .await?,
-                Some(CapGrant::ChainAuthor(alice.clone())),
-            );
+        // alice should find her own authorship with higher priority than the
+        // committed grant even if she passes in the secret
+        assert_eq!(
+            chain
+                .valid_cap_grant(function.clone(), alice.clone(), secret.clone())
+                .await?,
+            Some(CapGrant::ChainAuthor(alice.clone())),
+        );
 
-            // bob should be granted with the committed grant as it matches the secret he passes to
-            // alice at runtime
-            assert_eq!(
-                chain
-                    .valid_cap_grant(function.clone(), bob.clone(), secret.clone())
-                    .await?,
-                Some(grant.clone().into())
-            );
-        }
+        // bob should be authorized with transferable cap grant when passing in the secret
+        assert_eq!(
+            chain
+                .valid_cap_grant(function.clone(), bob.clone(), secret.clone())
+                .await?,
+            Some(grant.clone().into())
+        );
+        // bob should not be authorized for other zomes/functions than the ones granted
+        assert_eq!(
+            chain
+                .valid_cap_grant(("boo".into(), "far".into()), bob.clone(), secret.clone())
+                .await?,
+            None
+        );
 
-        // let's roll the secret and assign the grant to bob specifically
+        // convert transferable cap grant to assigned with bob as assignee
         let mut assignees = BTreeSet::new();
         assignees.insert(bob.clone());
         let updated_secret = Some(CapSecretFixturator::new(Unpredictable).next().unwrap());
         let updated_access = CapAccess::from((updated_secret.clone().unwrap(), assignees));
         let updated_grant = ZomeCallCapGrant::new("tag".into(), updated_access.clone(), functions);
 
+        // commit grant update to alice's source chain
         let (updated_action_hash, updated_entry_hash) = {
             let chain = SourceChain::new(
```

Comment:
Flushing the SourceChain doesn't actually update its notion of current head, because it's intended to be dropped after the flush, and recreating it is what updates the current head. It would actually be better if `flush` took `self` instead of `&self` for this reason. You were able to remove the other one because there were no new commits on that one, only reads.